engine/resources/hetzner.go

```
@@ -0,0 +1,710 @@
+func (obj *HetznerRes) Cmp(r engine.Res) error {
+	// check if empty
+	if obj == nil && r == nil {
+		return nil
+	}
+	if (obj == nil) != (r == nil) {
+		return fmt.Errorf("one resource is empty")
+	}
+	// compare types
+	res, ok := r.(*HetznerRes)
+	if !ok {
+		return fmt.Errorf("not a %s", obj.Kind())
+	}
+	// compare resource fields
+	if obj.State != res.State {
+		return fmt.Errorf("state differs")
+	}
+	if obj.APIToken != res.APIToken {
+		return fmt.Errorf("apitoken differs")
+	}
+	if obj.Servertype != res.Servertype {
+		return fmt.Errorf("servertype differs")
+	}
+	if obj.Datacenter != res.Datacenter {
+		return fmt.Errorf("datacenter differs")
+	}
+	if obj.Image != res.Image {
+		return fmt.Errorf("image differs")
+	}
+	if obj.ServerRescueMode != res.ServerRescueMode {
+		return fmt.Errorf("serverrescuemode differs")
+	}
+	for i, key := range obj.ServerRescueSSHKeys {
+		if key != res.ServerRescueSSHKeys[i] {
+			return fmt.Errorf("serverrescuekeys differs")
+		}
+	}
+	return nil
+}
+
+// cmpServerSpecs compares the server specifications between the virtual
+// HetznerRes resource and its corresponding Hetzner instance.
+// Returns true if Servertype, Datacenter and Image match.
+func (obj HetznerRes) cmpServerSpecs() (bool, error) {
+	if obj.init.Debug {
+		obj.init.Logf("cmpServerSpecs()")
+	}
+	if obj.server == nil {
+		return false, fmt.Errorf("server is unavailable")
+	}
+	if obj.Servertype != obj.server.ServerType.Name {
+		return false, nil
+	}
+	if obj.Datacenter != obj.server.Datacenter.Name {
+		return false, nil
+	}
+	if obj.Image != obj.server.Image.Name {
+		return false, nil
+	}
+	return true, nil
+}
+
+// stateConverged checks if the target server is in the desired state.
+// Returns true if both values agree that the server is powered on or off,
+// or "absent" as intended. Returns false otherwise.
+func (obj HetznerRes) stateConverged() bool {
+	if obj.init.Debug {
+		obj.init.Logf("stateConverged()")
+	}
+	var statesMatch bool
```

https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow
(or return flow)
basically programmers always want to avoid elseif when we can return early unless absolutely necessary

---

I fixed the corresponding TODO to throw an error in case of unknown keynames, rather than silently ignoring them. Seems like the safer way to handle those exceptions, and now the code looks much cleaner too.